pkg/cri/sbserver/sandbox_status.go

```diff
@@ -42,12 +43,32 @@ func (c *criService) PodSandboxStatus(ctx context.Context, r *runtime.PodSandbox
 		return nil, fmt.Errorf("failed to get sandbox controller: %w", err)
 	}
 
+	var (
+		createdAt time.Time
+		state     string
+		info      map[string]string
+	)
 	cstatus, err := controller.Status(ctx, sandbox.ID, r.GetVerbose())
 	if err != nil {
-		return nil, fmt.Errorf("failed to query controller status: %w", err)
+		// If the shim died unexpectedly (segfault etc.) let's set the state as
+		// NOTREADY and not just error out to make k8s and clients like crictl
+		// happy. If we get back ErrNotFound from controller.Status above while
+		// we're using the shim-mode controller, this is a decent indicator it
+		// exited unexpectedly. We can use the fact that we successfully retrieved
+		// the sandbox object from the store above to tell that this is true, otherwise
+		// if we followed the normal k8s convention of StopPodSandbox -> RemovePodSandbox,
+		// we wouldn't have that object in the store anymore.
+		if !errdefs.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to query controller status: %w", err)
+		}
+		state = runtime.PodSandboxState_SANDBOX_NOTREADY.String()
```

Comment:
hmm.. for consideration to make sure the corner cases are covered:
https://github.com/containerd/containerd/blob/main/pkg/cri/store/sandbox/status.go#L70-L75

if adding another case to the state machine, such as changing the declaration ^ that NOTREADY "ONLY" means the sandbox container is not running we should at least update the doc there and check if we have integration / e2e buckets for the new case, sorry if those old docs state machine is hard to read :-)

Reply:
Mike and I just spoke offline, this should be fine but we need more docs for the Sandbox API on what is a valid State string to return here. The sandbox api isn't tied to CRI (although it maps fairly nice as we've seen) but currently this code expects the string returned to be the string representation of k8s' valid pod state values (SANDBOX_READY or SANDBOX_NOTREADY). In defense of using NOTREADY here (k8s only has two values for a pods state so there's not much else we can use), we actually treat IsNotFound the same way when loading sandboxes after a restart already https://github.com/containerd/containerd/blob/main/pkg/cri/sbserver/restart.go#L107-L112

Reply:
The return values are fixed by cri because k8s had a limited set of known states for a pod (2).. the internal value also included unknown to cover the corner case specified where it is neither ready nor not ready. Need to understand/document how we'll know the difference between the two versions of not ready proposed without using another state ..like unknown.

Reply:
a todo / issue covering either adding back in the unknown path.. or explaining how we handle both paths
start failed not cleaned or reload failed going down the unknown path

Reply:
I was following originally, but now I'm not. This is solely to appease k8s and make us able to make forward progress here. Do you agree if a shim dies unexpectedly we should be able to make forward progress until we get to calling Remove, as that's what the goal was here. For our own state tracking (where we have the three Ready/NotReady/Unknown states), if the shim dies, for our own CRI plugin specific state tracking this will trigger the sandbox APIs Wait() method to return which eventually leads us to setting the sandbox state to NotReady as well after looking: https://github.com/containerd/containerd/blob/290a800e83d5460207cf198516359bfd2b5038d6/pkg/cri/server/events.go#L419-L437. So our own state is set to NOTREADY, and we send to k8s NOTREADY as well, so we're both in agreeance actually after looking deeper

> a todo / issue covering either adding back in the unknown path..

What unknown path was there here that we've gotten rid of?

Reply:
mostly it's start failed and we have not cleaned the mess up yet, but also the case where there is a restart while in that state/ load failed 

I see there is already one todo for this https://github.com/containerd/containerd/blob/main/pkg/cri/sbserver/podsandbox/sandbox_stop.go#L46